sdks/go/pkg/beam/core/runtime/harness/harness.go

```diff
@@ -273,6 +273,54 @@ func (c *control) getOrCreatePlan(bdID bundleDescriptorID) (*exec.Plan, error) {
 	return plan, nil
 }
 
+// Sampler provides methods for implementing state sampling
+// to track execution time metrics.
+type Sampler interface {
+	startSampler()
+	stopSampler()
+	sample()
+}
+
+// StateSampler tracks the state of a bundle.
+type StateSampler struct {
+	done  chan (int) // signal to stop sampling
+	e     *metrics.ExecutionStateTracker
+	store *metrics.Store
+	pid   string // PTransform ID
+}
+
+func newSampler(e *metrics.ExecutionStateTracker, store *metrics.Store, pid string) StateSampler {
+	return StateSampler{done: make(chan int), e: e, store: store, pid: pid}
+}
+
+func (s StateSampler) startSampler() {
+	for {
+		select {
+		case <-s.done:
+			return
+		default:
+			if s.e.NumberOfTransitions != s.e.TransitionsAtLastSample {
+				s.sample()
+			}
+			s.e.MillisSinceLastTransition += 200
+			s.e.State.TotalTimeMillis += 200
+			time.Sleep(200 * time.Millisecond)
```

Comment:
The harness package should be what handles the times/sleeps however. We can have the tick/increment method on the  executions state tracker take in the number of milliseconds that have passed. That will make it easier to test by itself (avoiding real time in tests).

To avoid lock contention or similar, this implies we might need the tick method, and a "sampleAndTick" method to avoid too much lock churn.